### ISSUE TITLE
VA-API support breaks app

### ISSUE THREAD

**Issue by JoelleJS** [flathub/org.kde.krdc#81]
### Issue
When connecting to an RDP server the screen stays blank and I get the following error repeatedly:
```
[16:09:36:001] [2:50] [ERROR][com.freerdp.codec] - Failed to decode video frame (status=-1094995529)
[16:09:36:001] [2:50] [WARN][com.freerdp.gdi] - avc420_decompress failure: -1, ignoring update.
[h264 @ 0x7c2fac12e7c0] Hardware does not support image size 4480x1440 (constraints: width 64-4096 height 64-4096).
[h264 @ 0x7c2fac12e7c0] Failed setup for format vaapi: hwaccel initialisation returned error.
[h264 @ 0x7c2fac12e7c0] decode_slice_header error
[h264 @ 0x7c2fac12e7c0] Hardware does not support image size 4480x1440 (constraints: width 64-4096 height 64-4096).
[h264 @ 0x7c2fac12e7c0] Failed setup for format vaapi: hwaccel initialisation returned error.
[h264 @ 0x7c2fac12e7c0] decode_slice_header error
[h264 @ 0x7c2fac12e7c0] Hardware does not support image size 4480x1440 (constraints: width 64-4096 height 64-4096).
[h264 @ 0x7c2fac12e7c0] Failed setup for format vaapi: hwaccel initialisation returned error.
[h264 @ 0x7c2fac12e7c0] decode_slice_header error
[h264 @ 0x7c2fac12e7c0] Hardware does not support image size 4480x1440 (constraints: width 64-4096 height 64-4096).
[h264 @ 0x7c2fac12e7c0] Failed setup for format vaapi: hwaccel initialisation returned error.
[h264 @ 0x7c2fac12e7c0] decode_slice_header error
[h264 @ 0x7c2fac12e7c0] Hardware does not support image size 4480x1440 (constraints: width 64-4096 height 64-4096).
[h264 @ 0x7c2fac12e7c0] Failed setup for format vaapi: hwaccel initialisation returned error.
[h264 @ 0x7c2fac12e7c0] decode_slice_header error
[h264 @ 0x7c2fac12e7c0] Hardware does not support image size 4480x1440 (constraints: width 64-4096 height 64-4096).
[h264 @ 0x7c2fac12e7c0] Failed setup for format vaapi: hwaccel initialisation returned error.
[h264 @ 0x7c2fac12e7c0] decode_slice_header error
[h264 @ 0x7c2fac12e7c0] Hardware does not support image size 4480x1440 (constraints: width 64-4096 height 64-4096).
[h264 @ 0x7c2fac12e7c0] Failed setup for format vaapi: hwaccel initialisation returned error.
[h264 @ 0x7c2fac12e7c0] decode_slice_header error
[h264 @ 0x7c2fac12e7c0] Hardware does not support image size 4480x1440 (constraints: width 64-4096 height 64-4096).
[h264 @ 0x7c2fac12e7c0] Failed setup for format vaapi: hwaccel initialisation returned error.
[h264 @ 0x7c2fac12e7c0] decode_slice_header error
[h264 @ 0x7c2fac12e7c0] no frame!
```

### Presumable cause
It seems that #42 has broken the app for me. I have no issues with the following:
* The Arch Linux package `krdc`
* The Flatpak when I build it myself with the line from #42 removed

I presume that it has to do with having an AMD GPU. #42 mentions:
> Requires the correct org.freedesktop.Platform.VAAPI Flatpak to be installed.

No such flatpak is available for AMD, there is only org.freedesktop.Platform.VAAPI.Intel.

### System information
OS: Arch Linux x86_64
Kernel: Linux 6.11.5-arch1-1
DE: KDE Plasma 6.2.2
WM: KWin (Wayland)
CPU: Intel(R) Core(TM) i7-7700K
GPU: AMD Radeon RX 5500 XT

**Comment by Pointedstick** [comment]
Can confirm.

**Comment by travier** [comment]
Should we revert https://github.com/flathub/org.kde.krdc/pull/42 for now?
This looks like an upstream KRDC or freerdp bug.

**Comment by tsdgeos** [comment]
If reverting it fixes it, sure

**Comment by JoelleJS** [comment]
I do wonder if there's another solution, I can find discussions of other Flatpak'ed apps using VA-API working on NVIDIA and AMD and only Intel requiring the extra `org.freedesktop.Platform.VAAPI.Intel`. E.g. https://discourse.flathub.org/t/how-to-enable-video-hardware-acceleration-on-flatpak-firefox/3125. Makes me wonder if this should be dived in further. If we can have hardware acceleration for all that's a good thing.

**Comment by travier** [comment]
Made https://github.com/flathub/org.kde.krdc/pull/86

**Comment by JoelleJS** [comment]
I see it is merged, building and testing right now

**Comment by twelho** [comment]
@JoelleJS from the logs I see that your video stream is very wide (4480 pixels) and that exceeds the hardware decoder's capabilities. Can you try running with VAAPI, but make the desktop/window narrower, so it would fit within the 64-4096 pixel bounds? Does hardware acceleration work then?

**Comment by twelho** [comment]
Also, are you using a hardware acceleration flag such as `/gfx-h264:avc444`? IIRC FreeRDP shouldn't try to use hardware acceleration by default, unless you explicitly request it. At least that was the case for me, but I don't have a test setup to confirm that right now.

**Comment by JoelleJS** [comment]
@twelho Ah yes, that's KRDP combining my two displays into one very wide RDP stream. No luck when I disable one of my screens though:
```
flatpak run --system app/org.kde.krdc/x86_64/stable
KRDC: Starting RDP session
[15:06:15:023] [2:2] [WARN][com.freerdp.crypto] - Certificate verification failure 'self-signed certificate (18)' at stack position 0
[15:06:15:023] [2:2] [WARN][com.freerdp.crypto] - C = AU, ST = Some-State, O = Internet Widgits Pty Ltd
[15:06:20:203] [2:2] [INFO][com.freerdp.gdi] - Local framebuffer format  PIXEL_FORMAT_RGBX32
[15:06:20:203] [2:2] [INFO][com.freerdp.gdi] - Remote framebuffer format PIXEL_FORMAT_BGRA32
[15:06:20:240] [2:2] [INFO][com.freerdp.channels.rdpsnd.client] - [static] Loaded pulse backend for rdpsnd
[15:06:20:240] [2:2] [INFO][com.freerdp.channels.drdynvc.client] - Loading Dynamic Virtual Channel rdpsnd
[15:06:20:240] [2:2] [INFO][com.freerdp.channels.drdynvc.client] - Loading Dynamic Virtual Channel audin
[15:06:20:243] [2:2] [INFO][com.freerdp.channels.audin.client] - Loaded pulse backend for audin
[15:06:20:243] [2:2] [INFO][com.freerdp.channels.drdynvc.client] - Loading Dynamic Virtual Channel rdpgfx
[15:06:20:351] [2:58] [ERROR][com.freerdp.codec] - Failed to transfer video frame (status=-22) (Invalid argument)
[15:06:20:351] [2:58] [WARN][com.freerdp.gdi] - avc420_decompress failure: -1, ignoring update.
[15:06:20:364] [2:58] [ERROR][com.freerdp.codec] - Failed to transfer video frame (status=-22) (Invalid argument)
[15:06:20:364] [2:58] [WARN][com.freerdp.gdi] - avc420_decompress failure: -1, ignoring update.
[15:06:20:380] [2:58] [ERROR][com.freerdp.codec] - Failed to transfer video frame (status=-22) (Invalid argument)
[15:06:20:380] [2:58] [WARN][com.freerdp.gdi] - avc420_decompress failure: -1, ignoring update.
[15:06:20:412] [2:58] [ERROR][com.freerdp.codec] - Failed to transfer video frame (status=-22) (Invalid argument)
[15:06:20:412] [2:58] [WARN][com.freerdp.gdi] - avc420_decompress failure: -1, ignoring update.
[15:06:20:435] [2:58] [ERROR][com.freerdp.codec] - Failed to transfer video frame (status=-22) (Invalid argument)
[15:06:20:435] [2:58] [WARN][com.freerdp.gdi] - avc420_decompress failure: -1, ignoring update.
[15:06:20:463] [2:58] [ERROR][com.freerdp.codec] - Failed to transfer video frame (status=-22) (Invalid argument)
[15:06:20:463] [2:58] [WARN][com.freerdp.gdi] - avc420_decompress failure: -1, ignoring update.
[15:06:20:593] [2:58] [ERROR][com.freerdp.codec] - Failed to transfer video frame (status=-22) (Invalid argument)
[15:06:20:593] [2:58] [WARN][com.freerdp.gdi] - avc420_decompress failure: -1, ignoring update.
[15:06:20:630] [2:58] [ERROR][com.freerdp.codec] - Failed to transfer video frame (status=-22) (Invalid argument)
[15:06:20:630] [2:58] [WARN][com.freerdp.gdi] - avc420_decompress failure: -1, ignoring update.
[15:06:20:659] [2:58] [ERROR][com.freerdp.codec] - Failed to transfer video frame (status=-22) (Invalid argument)
[15:06:20:659] [2:58] [WARN][com.freerdp.gdi] - avc420_decompress failure: -1, ignoring update.
[15:06:20:667] [2:59] [INFO][com.freerdp.core] - ERRINFO_LOGOFF_BY_USER (0x0000000C):The disconnection was initiated by the user logging off their session on the server.
[15:06:20:667] [2:59] [ERROR][com.freerdp.core] - rdp_set_error_info:freerdp_set_last_error_ex ERRINFO_LOGOFF_BY_USER [0x0001000C]
[15:06:20:667] [2:59] [ERROR][com.freerdp.core.rdp] - rdp_recv_tpkt_pdu: rdp_recv_deactivate_all() fail
[15:06:20:667] [2:59] [ERROR][com.freerdp.core.transport] - transport_check_fds: transport->ReceiveCallback() - -1
```

**Comment by JoelleJS** [comment]
Can confirm that it is working after #86

**Comment by twelho** [comment]
> @twelho Ah yes, that's KRDP combining my two displays into one very wide RDP stream. No luck when I disable one of my screens though

Alright, then it makes sense to revert the HW acceleration for the general case. Just out of curiosity if we can get this to work at all, could you try adding either `/gfx-h264:avc444` or `/gfx-h264:avc420` to the flags and seeing if it works with one monitor?

**Comment by twelho** [comment]
KRDC uses FreeRDP under the hood to connect to RDP servers. Let me check how to add the flags again, it's been a little while...

**Comment by twelho** [comment]
Okay, so multiple things have happened since I last looked at this:

- KRDC used to launch FreeRDP as a separate process (separate window). However, it has been refactored to use FreeRDP as a library now, so you can [no longer specify FreeRDP flags](https://discuss.kde.org/t/advanced-settings-removed-from-krdc-in-v24-01-80/11080). Any options would need to be exposed in the GUI directly, and there is no "enable/disable hardware acceleration" toggle yet from what I can tell.
- FreeRDP is basically all-or-nothing wrt. VAAPI hardware acceleration, if you compile with `-DWITH_VAAPI:BOOL=ON` it will [always try to use VAAPI](https://github.com/FreeRDP/FreeRDP/blob/782974a3e40001af07213449400aee9083242797/libfreerdp/codec/h264_ffmpeg.c#L626-L651), especially since now the `/gfx-h264:` etc. flags are [deprecated](https://github.com/FreeRDP/FreeRDP/blob/782974a3e40001af07213449400aee9083242797/client/common/cmdline.c#L4951) in favor of the generic (previously unaccelerated IIRC) `/gfx:...` flags.
- Finally, while FreeRDP will do some functional tests for VAAPI (which succeed in your case, so AMD GPUs only need `org.freedesktop.Platform.VAAPI`), FreeRDP doesn't fall back to software rendering if the video stream is not hardware-decodable. I found [this exact issue with an out-of-bounds resolution](https://github.com/FreeRDP/FreeRDP/issues/8373) as an open bug in FreeRDP's issues.

TL;DR: FreeRDP doesn't handle hardware acceleration fallbacks gracefully enough yet to warrant it being enabled by default.

**Comment by JoelleJS** [comment]
Later today or tomorrow I will try compiling FreeRDP 2.11.7 and 3.10.2 with WITH_VAAPI=on and see if I can get either to work. That should give some insight.

**Comment by JoelleJS** [comment]
So I did those steps and both versions work on my system and are hardware accelerated. Since I'm on Arch Linux I figured that different library versions may be the problem. So I changed the version of `org.freedesktop.Platform.ffmpeg-full` in the manifest from 23.08 to 24.08, while also re-adding `WITH_VAAPI:BOOL=ON`. And guess what: it works without error and with hardware acceleration.

Now, this does not solve the third point @twelho raised. Also, I have not tested this on a system that doesn't support VAAPI, or a system that does support VAAPI but not the specific codec used by the RDP server you're connecting to. Also, VAAPI support is marked as `[experimental]` even in version 3.10.2 of FreeRDP, according to the logs I get.

It does make me wonder if the library version of FreeRDP could be a Flatpak extension with a VAAPI-fied and a non-VAAPI-fied version. That could complicate the installation of KRDP as a Flatpak though. Also, I don't know if Flatpak allows for mutually exclusive extensions that supply the same library.

Ideally this should be fixed in FreeRDP though.
